standalone/examples_test.go

```diff
@@ -0,0 +1,200 @@
+package standalone
+
+import (
+	"encoding/json"
+	"github.com/fullstorydev/grpcui/testing/testdata"
+	"github.com/google/go-cmp/cmp"
+	"testing"
+	"time"
+)
+
+func TestRequest_MarshalUnmarshal(t *testing.T) {
+	tests := []struct {
+		name    string
+		request Request
+	}{
+		{
+			name: "Full request, complex data",
+			request: Request{
+				Timeout: 1 * time.Second,
+				Metadata: []MetadataPair{
+					{
+						"key",
+						"value",
+					},
+				},
+				Data: map[string]interface{}{
+					"field": "fieldValue",
+				},
+			},
+		},
+		{
+			name: "Full request, string data",
+			request: Request{
+				Timeout: 2 * time.Millisecond,
+				Metadata: []MetadataPair{
+					{
+						"key",
+						"value",
+					},
+				},
+				Data: "a string",
+			},
+		},
+		{
+			name: "No timeout",
+			request: Request{
+				Metadata: []MetadataPair{
+					{
+						"key",
+						"value",
+					},
+				},
+				Data: map[string]interface{}{
+					"field": "fieldValue",
+				},
+			},
+		},
+		{
+			name: "no data",
+			request: Request{
+				Timeout: 1 * time.Millisecond,
+				Metadata: []MetadataPair{
+					{
+						"key",
+						"value",
+					},
+				},
+			},
+		},
+		{
+			name: "only timeout",
+			request: Request{
+				Timeout: 24 * time.Hour,
+			},
+		},
+		{
+			name: "only data",
+			request: Request{
+				Data: "some data",
+			},
+		},
+		{
+			name: "only metadata",
+			request: Request{
+				Metadata: []MetadataPair{
+					{
+						"a",
+						"b",
+					},
+				},
+			},
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			marshal, err := json.Marshal(test.request)
+			if err != nil {
+				t.Fatalf("marshal failed: %v", err)
+			}
+			t.Logf("marshaled: %q", string(marshal))
+
+			var unmarshaled Request
+			err = json.Unmarshal(marshal, &unmarshaled)
+			if err != nil {
+				t.Fatalf("unmarshal failed: %v", err)
+			}
+
+			if diff := cmp.Diff(test.request, unmarshaled); diff != "" {
+				t.Fatalf("OSRM gateway request mismatch (-want +got):\n%s", diff)
+			}
+		})
+	}
+}
+
+func TestRequest_MarshalJSON_ProtoData(t *testing.T) {
+	tests := []struct {
+		name    string
+		request Request
+	}{
+		{
+			name: "proto3 data",
+			request: Request{
+				Data: testmodels.TestMessage3{
+					AInt32:  107,
+					AString: "string",
+				},
+			},
+		},
+		{
+			name: "proto2 data",
+			request: Request{
+				Data: testmodels.TestMessage2{
+					AInt32:  int32Ptr(107),
+					AString: strPtr("string"),
+				},
+			},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			marshal, err := json.Marshal(test.request)
+			if err != nil {
+				t.Fatalf("marshal failed: %v", err)
+			}
+			t.Logf("marshaled: %q", string(marshal))
+
+			var raw interface{}
```

Comment:
If you unmarshal into a custom struct, you don't have to do the gross checks below but can instead just use `proto.Equal` to compare the unmarshaled proto to the original.

Reply:
Not sure I understand how we can decode a `proto.Message` in `ExampleRequest.Data` as implemented now, `Data` will contain whatever json.Decode returned. We'd need to add an extra step to convert from `interface{}` to `proto.Message`. Can you please give an example?

Reply:
You currently aren't using `ExampleRequest`, so would continue to do that. So instead of unmarshaling into `interface{}`, you'd unmarshal into a 
```go
// this happens to exactly match the struct in the main `github.com/fullstorydev/grpcui` package
// that is used to parse request that is sent from the browser...
type request struct {
    TimeoutSeconds float32 `json:"timeout_seconds"`
    Metadata []ExampleMetadataEntry `json:"metadata"`
    Data json.RawMessage `json:"data"`
}
// You can then use jsonpb to unmarshal req.Data into whatever protobuf message and
// then use proto.Equal to verify that the unmarshaled proto matches the one that was
// originally marshaled.
```

Reply:
updated tests